### PR TITLE
Do not pull existing local images.

### PR DESCRIPTION
**Pull request by shado1111w**

### CODE REVIEW

File: docker.go [hunk 1]
```diff
@@ -58,6 +60,29 @@ func (g *g) dockerConnect() (*docker, error) {
...
+	images, err := d.client.ImageList(ctx, types.ImageListOptions{All: true})
+	if err != nil {
+		return false, fmt.Errorf("can't list image: %w", err)
+	}
+
+	for _, img := range images {
+		for _, repoTag := range img.RepoTags {
+			if strings.HasSuffix(image, repoTag) {
+				return true, nil
+			}
```
Comment (orlangure) [inline docker.go:79]: The problem you are describing in the comments relates to short image names, without docker organization. From my experience so far, if an image has a short name, like `mongo`, it is safe to convert it to `library/mongo`. If we add the following piece of code just before these 3 lines, the whole edge case of `mongo` vs `circleci/mongo` should go away:

```go
			if !strings.Contains(repoTag, "/") {
				repoTag = "library/" + repoTag
			}
```

Making this change will allow to reduce the complexity of the code by removing the need for an optional retry that you added to mitigate false-positives, regular expression matches and probably the entire `prepareContainer` function.

The good thing is that you added a test that covers this edge case, and we will know if my suggested change breaks anything or not.

File: gnomock_test.go [hunk 2]
```diff
@@ -192,6 +192,42 @@ func TestGnomock_withCommand(t *testing.T) {
...
+		mongoImage         = "docker.io/library/mongo:4.4"
+		circleciMongoImage = "docker.io/circleci/mongo:4.4"
+	)
+
+	container, err := gnomock.StartCustom(
+		mongoImage,
+		gnomock.DefaultTCP(testutil.GoodPort80),
+		gnomock.WithUseLocalImagesFirst(),
+	)
+	require.Nil(t, err)
```
Comment (orlangure) [inline gnomock_test.go:208]: The result would be the same, but I personally prefer to use `require.NoError` as it producer more clear test failure output ("unexpected error").

File: options.go [hunk 3]
```diff
@@ -154,6 +154,13 @@ func WithDisableAutoCleanup() Option {
 	}
 }
 
+// WithUseLocalImagesFirst will try to use existing local image rather than always pull the image.
```
Comment (orlangure) [inline options.go:157]:
```
 // WithUseLocalImagesFirst allows to use existing local images if possible
 // instead of always pulling the images.
```

File: options.go [hunk 4]
```diff
@@ -224,6 +231,9 @@ type Options struct {
 	// will try to stop containers created by it right after the tests exit.
 	DisableAutoCleanup bool `json:"disable_cleanup"`
 
+	// UseLocalImagesFirst to avoid hitting the docker hub pull rate limit.
```
Comment (orlangure) [inline options.go:234]:
```suggestion
	// UseLocalImagesFirst if possible to avoid hitting the Docker Hub pull rate limit.
```

File: options.go [hunk 5]
```diff
@@ -224,6 +231,9 @@ type Options struct {
...
+	UseLocalImagesFirst bool `json:"use_local_images_first"`
```
Comment (orlangure) [inline options.go:235]: This option should be added to [Swagger spec](https://github.com/orlangure/gnomock/blob/master/swagger/swagger.yaml#L441) as well for non-Go usages.

**Review by orlangure — COMMENTED** [review]
Thank you for taking care of this so quickly!

I think I found a way to resolve the edge case you discovered and reduce the complexity of this change. Let me know what you think.

File: docker.go (hunk 1 above)
Comment (shado1111w) [reply docker.go:79]: Thanks. It is the better way to resolve the edge case.

**Review by orlangure — APPROVED** [review]
Cool, thank you for contribution! 😻
